### PR TITLE
Alerting: Add (manual) script to generate alerting RTKQ APIs from OpenAPI spec

### PR DESCRIPTION
**Pull request by tomratcliffe**
**What is this feature?**

Generates RTK Query API client for the UI to use, based on the OpenAPI spec from https://github.com/grafana/grafana/pull/88201

**Why do we need this feature?**

Generating the RTKQ API directly from the OpenAPI spec is a sensible thing to do as it takes a lot of the boilerplate out of the equation. As more APIs will be written in this way in the future, we should aim to make this part of the workflow for consuming new APIs

**Who is this feature for?**

Alerting UI devs

To generate the RTKQ API, the response from `/openapi/v3/apis/notifications.alerting.grafana.app/v0alpha1` needs to be saved to `data/alerting/openapi.json`. This is because (to my understanding), the OpenAPI specs aren't all written to disk - some are generated on the fly based on the "apps" running on the instance

### CODE REVIEW

File: scripts/generate-alerting-rtk-apis.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+const config: ConfigFile = {
+  schemaFile,
+  apiFile: '',
+  tag: true,
+  outputFiles: {
+    '../public/app/features/alerting/unified/openapi/timeIntervalsApi.gen.ts': {
+      apiFile: '../public/app/features/alerting/unified/api/alertingApi.ts',
+      apiImport: 'alertingApi',
+      filterEndpoints: [/TimeInterval/],
```
Comment (tomratcliffe) [inline scripts/generate-alerting-rtk-apis.ts:32]: I'm filtering to get all TimeInterval operations here, but do we want to explicitly pick some as we need them? And regenerate the RTKQ client as we add them in?

File: scripts/generate-alerting-rtk-apis.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  console.error(
+    'Please visit /openapi/v3/apis/notifications.alerting.grafana.app/v0alpha1 and save the OpenAPI definition to data/alerting/openapi.json'
+  );
+  process.exit(1);
+}
+
+const config: ConfigFile = {
+  schemaFile,
+  apiFile: '',
+  tag: true,
```
Comment (tomratcliffe) [inline scripts/generate-alerting-rtk-apis.ts:27]: This adds the `provides`/`invalidates` tags properties - we can see how accurately this works once we start consuming this API

File: scripts/generate-alerting-rtk-apis.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+import { ConfigFile } from '@rtk-query/codegen-openapi';
+import fs from 'fs';
+
+const schemaFile = '../data/alerting/openapi.json';
+
+try {
+  // Check we have the OpenAPI before generating alerting RTK APIs,
+  // as this is currently a manual process
```
Comment (tomratcliffe) [inline scripts/generate-alerting-rtk-apis.ts:13]: As mentioned in the description, we currently have to manually save the OpenAPI json to the data folder and generate from there 🤷

File: scripts/generate-alerting-rtk-apis.ts (hunk 1 above)
Comment (tomratcliffe) [reply scripts/generate-alerting-rtk-apis.ts:32]: (I'm leaning towards explicitly including the ones we use)

File: scripts/generate-alerting-rtk-apis.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  fs.existsSync(schemaFile);
```
[the remark this answers is not in the document]
Comment (tomratcliffe) [reply scripts/generate-alerting-rtk-apis.ts]: I've updated it to use `accessSync` - but all I could see was that `exists` was deprecated, but `existsSync` wasn't. But regardless - using `access` now :+1:

File: public/app/features/alerting/unified/api/timeIntervalsApi.ts
Comment (tomratcliffe) [inline public/app/features/alerting/unified/api/timeIntervalsApi.ts:1]: (I might remove this entirely in followup PRs and instead create hooks that implement the endpoints under the hood. It's feeling easier to manage the types in that manner, for now)

File: scripts/generate-alerting-rtk-apis.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+const config: ConfigFile = {
+  schemaFile,
+  apiFile: '',
+  tag: true,
+  outputFiles: {
+    '../public/app/features/alerting/unified/openapi/timeIntervalsApi.gen.ts': {
+      apiFile: '../public/app/features/alerting/unified/api/alertingApi.ts',
+      apiImport: 'alertingApi',
+      filterEndpoints: ['listTimeIntervalForAllNamespaces'],
```
Comment (tomratcliffe) [inline scripts/generate-alerting-rtk-apis.ts:31]: Switched to explicitly mentioning the endpoints here, as it
* makes it clearer when a given PR adds a reliance on a new endpoint
* we end up with less code that's potentially unused
* searching through the codebase is more likely to bring up this file, leading to better understanding of how the code is created

File: public/app/features/alerting/unified/openapi/timeIntervalsApi.gen.ts
Comment (tomratcliffe) [inline public/app/features/alerting/unified/openapi/timeIntervalsApi.gen.ts:1]: (This is all generated from the OpenAPI spec, so no real need to review it/I can't change this one anyway, we would override it elsewhere)

File: scripts/generate-alerting-rtk-apis.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+/**
+ * To generate alerting k8s APIs, run:
+ * `npx rtk-query-codegen-openapi ./scripts/generate-alerting-rtk-apis.ts`
+ */
+
```
Comment (soniaAguilarPeiron) [inline scripts/generate-alerting-rtk-apis.ts:5]: maybe we should consider creating a new folder for including all the `generate-*-rtk-apis.ts`? I see that so far we are only two, but probably we will be more.
for example `scripts/rtk-gen/alerting/ .....`

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM 🚀! I understand we are going to iterate over it, and for now it's a manual script.
I also added a non-blocking suggestion.

**Comment by tomratcliffe** [comment]
Thanks Sonia! I'll leave it where it is for now, as we'll have to likely iterate on the script as we go anyway, as we add other APIs and try and automate this process?

**Pull request merged**